crates/walrus-service/src/node/storage/blob_info.rs

```diff
@@ -913,33 +1010,40 @@ impl ValidBlobInfoV1 {
     }
 
     fn maybe_increase_latest_deletable_registered_epoch(&mut self, epoch: Epoch) {
-        self.latest_seen_deletable_registered_epoch = Some(
+        self.latest_seen_deletable_registered_end_epoch = Some(
             epoch.max(
-                self.latest_seen_deletable_registered_epoch
+                self.latest_seen_deletable_registered_end_epoch
                     .unwrap_or_default(),
             ),
         )
     }
 
     fn maybe_increase_latest_deletable_certified_epoch(&mut self, epoch: Epoch) {
-        self.latest_seen_deletable_certified_epoch = Some(
+        self.latest_seen_deletable_certified_end_epoch = Some(
             epoch.max(
-                self.latest_seen_deletable_certified_epoch
+                self.latest_seen_deletable_certified_end_epoch
                     .unwrap_or_default(),
             ),
         )
     }
 
-    /// Decrements a counter on blob deletion.
+    /// Decrements a counter on blob deletion and unsets the corresponding latest seen end epoch in
+    /// case the counter reaches 0.
     ///
-    /// If the counter is 0, an error is logged in release builds and the function panics in dev
-    /// builds.
-    fn decrement_deletable_counter_on_deletion(counter: &mut u32) {
+    /// If the counter is already 0, an error is logged in release builds and the function panics in
+    /// dev builds.
+    fn decrement_deletable_counter_and_maybe_unset_latest_end_epoch(
+        counter: &mut u32,
+        latest_seen_end_epoch: &mut Option<Epoch>,
+    ) {
         debug_assert!(*counter > 0);
         *counter = counter.checked_sub(1).unwrap_or_else(|| {
             tracing::error!("attempt to delete blob when count was already 0");
             0
         });
+        if *counter == 0 {
+            *latest_seen_end_epoch = None;
+        }
```

Comment:
question: we don't cleanup the BlobInfo entry yet if deletable == 0 and there is no permanent blob, is that correct?

Reply:
Yes, that's correct. We don't actually clean up the aggregate blob info at all. This will be part of the next step, that is the actual data cleanup. Until then, we need to keep track of all blobs for which we potentially store metadata or slivers.